teps/0091-trusted-resources.md

```diff
@@ -0,0 +1,329 @@
+configured to be verified will have no change in performance.
+
+Tekton Chains will only need to make simple decisions over only locally
+available data so there will be no change to performance.
+
+
+## Design Details
+
+### Sign the Resources
+
+To sign the Resource, we should provide command line tools to help users for signing. The command line should be able to do the following steps:
+1. Read the Resource file, unmarshall it as a go object and calculate the sha of the
+json marshalled bytes.
+2. Use signing tools to sign the bytes and get the signature.
+3. Store the string encoded signature to Resource
+
+The sample signed file looks like this:
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  annotations:
+    tekton.dev/signature: MEQCIHhFC480mv7mh/6dxAlp/mvuvGXanuSghMsT+iBhWxt5AiBDvHv8sfKjJ3Ozrzvp+cjr30AOx1SPQDCcaJjlpAoVrA==
+  name: example-task
+spec:
+  steps:
+  - args:
+    - Hello World!
+    command:
+    - echo
+    image: ubuntu
+    name: echo
+    resources: {}
+```
+
+`tekton.dev/signature` is used to store the signature.
+
+The signed Resource can be installed directly on the kubernetes cluster or built as OCI bundle to be stored in the registry.
+
+### Verify the Resources
+
+The verification should be done in the Tekton Pipeline's admission webhook. A new configmap should be added to gate the verification
+code.
+
+### Webhook and Configuration
+
+Validating webhook can use configmap to allow users to config when the Resource fails the verification
+1) Directly fail the run;
+2) Not fail the run, and when Tekton Pipeline's dependent knative version support admission webhook warnings, return warning in `apis.FieldError`.
+3) Skip the validation;
+
+Configuration sample:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: config-trusted-resources
+  namespace: tekton-pipelines
+data:
+  onError: "stopAndFail"
+  cosign-pubkey-path: "/etc/signing-secrets/cosign.pub"
+  kms-pubkey-path: "gcpkms://projects/<project>/locations/<location>/keyRings/<keyring>/cryptoKeys/<key>"
+```
+`onError: [ stopAndFail | continue ]`. (Optional, default `continue`)
+`stopAndFail` will cause the build to fail if image verification fails,  `continue` will allow the build to continue even
+if verification fails.
+
+`skip-verification`. (Optional, default `true`), if true then directly return nil, if false then do the verfication
+
+`cosign-pubkey-path`. (Optional, default `/etc/signing-secrets/cosign.pub`), it specifies the secret path to store the cosign pubkey
+
+`kms-pubkey-path`.  (Optional, default empty), it specifies the KMS reference
+
+
+### Integrate with Remote Resource Resolution
+
+TEP-0060 introduces a new ResourceRequest reconciler that will actually pull
+the remote image and expand it to a Task.
+
+There are several ways to integrate the Trusted Tesource with Remote Resource Resolution
+
+
+| Method | Pros | Cons |
```

Comment:
Are you leaning towards one of these options at the moment?

Reply:
Right now the implementation is based on callbacks in validation webhook of pipeline. I think the benefit of having it in webhook is that we can fail the run/build instantly when it fails the verification.

I discussed the integration with remote resolution in one of the working groups before, and Priti suggested me to wait until the remote resolution come to beta version.